File: server/ruoyi-modules/jshpos-promotion/src/test/java/com/jingshanghui/pos/promotion/domain/ManualAdjustmentGoldenVectorTest.java
package com.jingshanghui.pos.promotion.domain;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.jingshanghui.pos.promotion.domain.ManualAdjustmentEngine.*;
import com.jingshanghui.pos.promotion.domain.PromotionModels.QuoteLine;
import com.jingshanghui.pos.promotion.domain.PromotionModels.QuoteResult;
import org.junit.jupiter.api.Test;

import java.math.BigDecimal;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

import static org.assertj.core.api.Assertions.assertThat;

/** Java 读取 PRM-002 Java/Dart 共用黄金向量。 */
class ManualAdjustmentGoldenVectorTest {
    @Test
    void matchesSharedManualVectorsExactly() throws Exception {
        Path path = Path.of("..", "..", "..", "contracts", "t2", "gate5a", "test-vectors",
            "manual-adjustment-vectors-v1.json").normalize();
        Map<String, Object> root = new ObjectMapper().readValue(Files.readString(path), new TypeReference<>() { });
        Map<String, Object> rawPolicy = map(root.get("policy"));
        Policy policy = new Policy(number(rawPolicy, "policyVersionId"), (String) rawPolicy.get("policySha256"),
            number(rawPolicy, "withoutApprovalMinor"), number(rawPolicy, "withApprovalMinor"),
            number(rawPolicy, "minimumLinePayableMinor"), number(rawPolicy, "maximumRoundingMinor"),
            list(rawPolicy.get("roundingMultiplesMinor")).stream().map(value -> ((Number) value).longValue()).toList());
        List<Map<String, Object>> rawLines = list(root.get("lines")).stream().map(this::map).toList();
        List<PromotionModels.BasketLine> lines = rawLines.stream().map(value -> new PromotionModels.BasketLine(
            (String) value.get("lineId"), ((Number) value.get("lineNo")).intValue(),
            Long.valueOf((String) value.get("skuId")), null, null,
            new BigDecimal((String) value.get("quantity")), number(value, "unitPriceMinor"))).toList();
        List<LineContext> contexts = lines.stream().map(value -> new LineContext(value.lineId(), value.lineNo(),
            value.skuId(), value.quantity())).toList();
        Map<String, Object> base = map(root.get("base"));
        Map<String, Long> baseDiscounts = longMap(base.get("lineDiscounts"));
        List<QuoteLine> quoteLines = lines.stream().map(value -> {
            long gross = value.quantity().multiply(BigDecimal.valueOf(value.unitPriceMinor())).longValueExact();
            return new QuoteLine(value.lineId(), gross, baseDiscounts.get(value.lineId()),
                gross - baseDiscounts.get(value.lineId()));
        }).toList();
        QuoteResult quote = new QuoteResult(number(base, "grossAmountMinor"), number(base, "discountAmountMinor"),
            number(base, "payableAmountMinor"), quoteLines, List.of(), List.of(), List.of());
        ManualAdjustmentEngine engine = new ManualAdjustmentEngine();
        for (Object raw : list(root.get("scenarios"))) {
            Map<String, Object> scenario = map(raw);
            Map<String, Object> expected = map(scenario.get("expected"));
            Preview result = engine.preview(quote, contexts, new Command("01K5R000000000000000000050",
                ActionType.valueOf((String) scenario.get("actionType")), (String) scenario.get("lineId"),
                (String) scenario.get("amountOrRate"), PaymentMethod.valueOf((String) scenario.get("paymentMethod"))), policy);
            assertThat(result.incrementalDiscountMinor()).as((String) scenario.get("id"))
                .isEqualTo(number(expected, "incrementalDiscountMinor"));
            assertThat(result.requiresApproval()).isEqualTo(expected.get("requiresApproval"));
            assertThat(result.result().payableAmountMinor()).isEqualTo(number(expected, "payableAmountMinor"));
            assertThat(result.result().lineDiscounts()).containsExactlyEntriesOf(longMap(expected.get("lineDiscounts")));
        }
    }

    @SuppressWarnings("unchecked") private Map<String, Object> map(Object value) { return (Map<String, Object>) value; }
    @SuppressWarnings("unchecked") private List<Object> list(Object value) { return (List<Object>) value; }
    private long number(Map<String, Object> value, String key) { return ((Number) value.get(key)).longValue(); }
    private Map<String, Long> longMap(Object value) {
        Map<String, Long> result = new LinkedHashMap<>();
        map(value).forEach((key, item) -> result.put(key, ((Number) item).longValue()));
        return result;
    }
}

File: pos-flutter/test/gate5a/manual_adjustment_golden_vector_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:flutter_test/flutter_test.dart';
import 'package:jshpos_pos/features/promotion/domain/manual_adjustment_engine.dart';
import 'package:jshpos_pos/features/promotion/domain/promotion_engine.dart';

void main() {
  test('Flutter matches every shared PRM-002 manual vector exactly', () {
    final root = jsonDecode(
      File(
        '../contracts/t2/gate5a/test-vectors/manual-adjustment-vectors-v1.json',
      ).readAsStringSync(),
    ) as Map<String, Object?>;
    final rawPolicy = root['policy']! as Map<String, Object?>;
    final policy = ManualPolicy(
      policyVersionId: rawPolicy['policyVersionId']! as int,
      policySha256: rawPolicy['policySha256']! as String,
      withoutApprovalMinor: rawPolicy['withoutApprovalMinor']! as int,
      withApprovalMinor: rawPolicy['withApprovalMinor']! as int,
      minimumLinePayableMinor: rawPolicy['minimumLinePayableMinor']! as int,
      maximumRoundingMinor: rawPolicy['maximumRoundingMinor']! as int,
      roundingMultiplesMinor: (rawPolicy['roundingMultiplesMinor']! as List)
          .cast<int>(),
    );
    final lines = (root['lines']! as List<Object?>).map((raw) {
      final value = raw! as Map<String, Object?>;
      return PromotionLine(
        lineId: value['lineId']! as String,
        lineNo: value['lineNo']! as int,
        skuId: value['skuId']! as String,
        categoryId: null,
        brandId: null,
        quantity: ExactDecimal.parse(
          value['quantity']! as String,
          maximumScale: 6,
        ),
        unitPriceMinor: value['unitPriceMinor']! as int,
      );
    }).toList();
    final rawBase = root['base']! as Map<String, Object?>;
    final base = PromotionQuote(
      grossAmountMinor: rawBase['grossAmountMinor']! as int,
      discountAmountMinor: rawBase['discountAmountMinor']! as int,
      payableAmountMinor: rawBase['payableAmountMinor']! as int,
      lineDiscounts: (rawBase['lineDiscounts']! as Map<String, Object?>).map(
        (key, value) => MapEntry(key, value! as int),
      ),
      appliedRuleIds: const [],
      explanations: const [],
      adjustments: const [],
    );
    final engine = ManualAdjustmentEngine();
    for (final raw in root['scenarios']! as List<Object?>) {
      final scenario = raw! as Map<String, Object?>;
      final expected = scenario['expected']! as Map<String, Object?>;
      final result = engine.preview(
        current: base,
        lines: lines,
        command: ManualCommand(
          authorizationId: '01K5R000000000000000000050',
          actionType: _action(scenario['actionType']! as String),
          lineId: scenario['lineId'] as String?,
          amountOrRate: scenario['amountOrRate']! as String,
          paymentMethod: scenario['paymentMethod'] == 'CASH'
              ? ManualPaymentMethod.cash
              : ManualPaymentMethod.nonCash,
        ),
        policy: policy,
      );
      expect(
        result.incrementalDiscountMinor,
        expected['incrementalDiscountMinor'],
        reason: scenario['id']! as String,
      );
      expect(result.requiresApproval, expected['requiresApproval']);
      expect(result.result.payableAmountMinor, expected['payableAmountMinor']);
      expect(
        result.result.lineDiscounts,
        (expected['lineDiscounts']! as Map<String, Object?>).map(
          (key, value) => MapEntry(key, value! as int),
        ),
      );
    }
  });
}

ManualActionType _action(String value) => switch (value) {
  'LINE_FIXED_PRICE' => ManualActionType.lineFixedPrice,
  'ORDER_AMOUNT_OFF' => ManualActionType.orderAmountOff,
  'ORDER_PERCENT_OFF' => ManualActionType.orderPercentOff,
  'ROUNDING' => ManualActionType.rounding,
  _ => throw FormatException('unknown action $value'),
};
